### PR TITLE
feat: add wrap_rank helper script

### PR DESCRIPTION
**Pull request by rb-determined-ai**
Core API users who use a hand-written or 3rd-party launcher will need to
prefix each log line with [rank=N] in order for the WebUI log viewer's
filter-by-rank feature to work.

Make this easier by exposing our internal wrapper script as a public
wrapper script.

The alternative of replacing sys.stdout with a class with a custom
write() method was deemed insufficient because many of our users depend
on C++ libraries that log information directly to the stdout file
descriptor.

### CODE REVIEW

File: docs/training-apis/api-core/distributed.txt [hunk 1]
```diff
@@ -21,25 +21,41 @@ considerations are:
 #. Create a ``4_distributed.py`` training script by copying the ``3_hpsearch.py`` from
    :ref:`core-hpsearch`.
 
-#. In your ``if __name__ == "__main__"`` block, add ``multiprocessing`` code to launch one process
-   per slot.
+#. Add launcher logic to execute one worker subprocess per slot.
+
+   Typically, your distributed training framework provides a launcher, as Horovod provides
```
Comment (gh-determined-ai) [inline docs/training-apis/api-core/distributed.txt:26]: I was never satisfied with this: maybe change it to "... **where** Horovod provides ..."

File: docs/training-apis/api-core/distributed.txt [hunk 2]
```diff
@@ -21,25 +21,41 @@ considerations are:
...
    .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.py
       :language: python
       :dedent:
-      :start-at: slots_per_node = len(info.slot_ids)
+      :pyobject: launcher_main
 
-   Typically, your distributed training framework provides a launcher, as Horovod provides
-   ``horovodrun`` and Torch provides ``torch.distributed.launch``, so you do not have to write the
-   launcher yourself.
+   Even when using a third-party provided launcher, you will still need to wrap your worker script
```
Comment (gh-determined-ai) [inline docs/training-apis/api-core/distributed.txt:37]: change to "... you need to wrap ..." [no future tense, no unnecessary words]

File: docs/training-apis/api-core/distributed.txt [hunk 3]
```diff
@@ -21,25 +21,41 @@ considerations are:
...
+   in the ``python -m determined.launcher.wrap_rank`` wrapper script for the WebUI log viewer to
+   filter logs by rank. This wrapper is automatically included when using the Determined-provided
+   launchers.
 
-   Even if the launcher is provided, you always need to create a
-   :class:`~determined.core.DistributedContext` on each worker to pass to ``core.init()``. In the
-   absence of a distributed training framework that might define rank/local_rank/cross_rank, you can
-   derive those values from the ClusterInfo API, as in this example.
+   Also add a ``worker_main()`` that will run training on each slot:
```
Comment (gh-determined-ai) [inline docs/training-apis/api-core/distributed.txt:42]: Change to: " Also, add a ``worker_main()`` function that runs training on each slot:"
[maybe 'method' is more accurate than 'function']

**Review by ioga — APPROVED** [review]

File: docs/training-apis/api-core/distributed.txt [hunk 4]
```diff
@@ -21,25 +21,41 @@ considerations are:
...
+   command-line arguments:
+
+   .. literalinclude:: ../../../examples/tutorials/core_api/4_distributed.py
+      :language: python
+      :dedent:
+      :start-at: slots_per_node = len(info.slot_ids)
 
-#. Use the ``allgather`` primitive to do a "distributed" increment, just to get a feel for using the
-   communication primitives:
+#. In the training code, use the ``allgather`` primitive to do a "distributed" increment, just to
```
Comment (gh-determined-ai) [inline docs/training-apis/api-core/distributed.txt:57]: Change "just to get a feel for using" to "to gain experience using"

File: harness/determined/launch/wrap_rank.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            "Can be an integer rank or the name of an "
+            "environment variable containing an integer rank."
+        ),
+    )
+    parser.add_argument(
+        "script", nargs=argparse.REMAINDER, metavar="SCRIPT...", help="The worker command."
+    )
+    args = parser.parse_args()
+
+    if set("0123456789") >= set(args.rank):
```
Comment (stoksc) [inline harness/determined/launch/wrap_rank.py:59]: lol, this is neat

**Review by gh-determined-ai — CHANGES_REQUESTED** [review]

File: harness/determined/launch/wrap_rank.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/launch/wrap_rank.py:59]: just recently learned this trick from @mpkouznetsov

File: docs/training-apis/api-core/distributed.txt (hunk 3 above)
Comment (rb-determined-ai) [reply docs/training-apis/api-core/distributed.txt:42]: There's no universal definition, but this [widely-upvoted stackoverflow post](https://stackoverflow.com/a/155633) reflects the definitions I am also most familiar with.

File: docs/training-apis/api-core/distributed.txt (hunk 1 above)
Comment (rb-determined-ai) [reply docs/training-apis/api-core/distributed.txt:26]: I totally rewrote this section.  I also didn't like the order of how I introduced ideas.  Please re-read.

**Review by gh-determined-ai — APPROVED** [review]
lgtm

**Pull request merged**